### PR TITLE
Fixes on cygwin and win32

### PR DESCRIPTION
**Pull request by pauloscustodio**
In cygwin, gcc -std=c++11 hides the realpath() function. Need to undefine __STRICT_ANSI__ to compile.
Need to include <climits> to have PATH_MAX.
Need to protect rvalue reference with #if / #endif for c++98 compiles.

### CODE REVIEW

File: filesystem/path.h [hunk 1]
```diff
@@ -18,6 +18,7 @@
 #include <cstdlib>
 #include <cerrno>
 #include <cstring>
+#include <climits>
```
Comment (Qix-) [inline filesystem/path.h]: What does this fix?

**Comment by pauloscustodio** [comment]
Without it I get on cygwin:

filesystem/path.h:83:19: error: ‘PATH_MAX’ was not declared in this scope
         char temp[PATH_MAX];

File: filesystem/path.h (hunk 1 above)
Comment (pauloscustodio) [inline filesystem/path.h]: Without it I get on cygwin:

filesystem/path.h:83:19: error: ‘PATH_MAX’ was not declared in this scope
         char temp[PATH_MAX];

File: CMakeLists.txt [hunk 2]
```diff
@@ -2,6 +2,6 @@ cmake_minimum_required(VERSION 2.8)
 project(path)
 if (CMAKE_CXX_COMPILER_ID MATCHES "Clang" OR CMAKE_CXX_COMPILER_ID MATCHES "GNU")
   # Enable C++11 mode on GCC / Clang
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -U__STRICT_ANSI__")
```
Comment (Qix-) [inline CMakeLists.txt]: Could you justify this a bit? I'm not entirely sure messing with system defines is the best solution here. What functionality is being blocked by this definition?

File: filesystem/path.h [hunk 3]
```diff
@@ -140,28 +145,44 @@ class path {
 
     std::string extension() const {
         const std::string &name = filename();
+        
+        if (name.empty() || name.front() == '.')
+            return "";
```
Comment (Qix-) [inline filesystem/path.h]:
```c++
return std::string();
```

File: filesystem/path.h [hunk 4]
```diff
@@ -140,28 +145,44 @@ class path {
...
+
         size_t pos = name.find_last_of(".");
         if (pos == std::string::npos)
             return "";
```
Comment (Qix-) [inline filesystem/path.h]:
```c++
return std::string();
```

File: filesystem/path.h [hunk 5]
```diff
@@ -184,49 +206,79 @@ class path {
...
+        if (m_type == windows_path && m_volume != 0)
+            oss << m_volume << ':';
+
+        if (m_absolute || m_starts_slash) {
+            oss << slash();
+            ends_slash = true;
+        }
 
-        for (size_t i=0; i<m_path.size(); ++i) {
+        for (size_t i = 0; i < m_path.size(); ++i) {
```
Comment (Qix-) [inline filesystem/path.h]:
```c++
for (size_t i = 0, path_size = m_path.size(); i < path_size; ++i) {
```

Probably a good idea to cache the path size here for posterity so you're not querying the size of the path each iteration. Can't guarantee the call will be optimized down to a simple lookup on certain platforms.

File: filesystem/path.h [hunk 6]
```diff
@@ -184,49 +206,79 @@ class path {
...
             oss << m_path[i];
+            ends_slash = false;
             if (i+1 < m_path.size()) {
```
Comment (Qix-) [inline filesystem/path.h]: Might want to cache here, too.

File: filesystem/path.h [hunk 7]
```diff
@@ -310,22 +362,29 @@ class path {
...
+                if (lastPos == std::string::npos || lastPos + 1 == string.length())
+                    break;
+                pos = string.find_first_of(delim, ++lastPos);
+            }
         }
 
         return tokens;
     }
 
+    char slash() const { return m_type == posix_path ? '/' : '\\'; }
```
Comment (Qix-) [inline filesystem/path.h]: Might want to break this out instead of having it on one line.

File: path_demo.cpp [hunk 8]
```diff
@@ -5,44 +5,373 @@
 using namespace std;
 using namespace filesystem;
 
-int main(int argc, char **argv) {
+// TAP-compliant test
```
Comment (Qix-) [inline path_demo.cpp]: The TAP compliant testing should be its own PR. Keep this PR to the scope of improving CYGWIN on Windows support.

Awesome that you got TAP testing working, though. :)

**Comment by pauloscustodio** [comment]
I have created a pull request for cygwin compile and one for TAP testing.
I will close this pull request and integrate the changes in another one dedicated to implement
similar behavior to boost::filesystem for the same methods.

File: path_demo.cpp (hunk 8 above)
Comment (pauloscustodio) [inline path_demo.cpp]: Thanks for the comments. I will take integrate them.
